### DISCUSSION TITLE
Motor Power Distribution from Thrust

### DISCUSSION THREAD

**Discussion by WoodyloveBo** [bitcraze/discussions#1320] [Q&A] [answered]
Hello.

I'm conducting an experiment on PWM to Thrust conversion for Crazyflie. While there are documents available from Bitcraze on this experiment, I'd like to proceed independently.

https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L85-L88

According to the link, the value obtained by adding motor.m1 to m4 and dividing by 4 should ultimately be the Thrust. To confirm this, I executed the following Python script, gradually increasing the Thrust from 10000.

```
import logging
import time
from threading import Thread
import csv

import cflib
from cflib.crazyflie import Crazyflie
from cflib.utils import uri_helper

from cflib.crazyflie.syncCrazyflie import SyncCrazyflie

from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncLogger import SyncLogger

uri = uri_helper.uri_from_env(default='radio://0/100/2M/E7E7E7E703')
logging.basicConfig(level=logging.ERROR)


class MotorRampExample:
    def __init__(self, link_uri):
        self._cf = Crazyflie(rw_cache='./cache')
        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)
        self._cf.connection_failed.add_callback(self._connection_failed)
        self._cf.connection_lost.add_callback(self._connection_lost)
        self._cf.open_link(link_uri)

        print('Connecting to %s' % link_uri)

    def _connected(self, link_uri):
        Thread(target=self._ramp_motors).start()

    def _connection_failed(self, link_uri, msg):
        print('Connection to %s failed: %s' % (link_uri, msg))

    def _connection_lost(self, link_uri, msg):
        print('Connection to %s lost: %s' % (link_uri, msg))

    def _disconnected(self, link_uri):
        print('Disconnected from %s' % link_uri)

    def _ramp_motors(self):
        thrust_mult = 1
        thrust_step = 500
        thrust = 10000
        pitch = 0
        roll = 0
        yawrate = 0

        # Unlock startup thrust protection
        self._cf.commander.send_setpoint(0, 0, 0, 0)

        logconf = LogConfig(name='MotorData', period_in_ms=100)
        logconf.add_variable('motor.m1', 'float')
        logconf.add_variable('motor.m2', 'float')
        logconf.add_variable('motor.m3', 'float')
        logconf.add_variable('motor.m4', 'float')
        logconf.add_variable('stabilizer.thrust', 'float')

        # Create and open CSV file for writing
        with open('motor_data2.csv', 'w', newline='') as csvfile:
            fieldnames = ['time', 'motor.m1', 'motor.m2', 'motor.m3', 'motor.m4', 'thrust']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()

            with SyncLogger(self._cf, logconf) as logger:
                for log_entry in logger:
                    data = log_entry[1]
                    m1_value = data['motor.m1']
                    m2_value = data['motor.m2']
                    m3_value = data['motor.m3']
                    m4_value = data['motor.m4']
                    thrust_value = data['stabilizer.thrust']

                    # Write data to CSV file
                    writer.writerow({
                        'time': log_entry[0],
                        'motor.m1': m1_value,
                        'motor.m2': m2_value,
                        'motor.m3': m3_value,
                        'motor.m4': m4_value,
                        'thrust': thrust_value
                    })

                    if thrust >= 25000:
                        thrust_mult = -1
                    thrust += thrust_step * thrust_mult
                    self._cf.commander.send_setpoint(roll, pitch, yawrate, thrust)
                    time.sleep(0.5) 

        self._cf.commander.send_setpoint(0, 0, 0, 0)
        time.sleep(0.1)
        self._cf.close_link()

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    example = MotorRampExample(uri)
```

(motor.m1+motor.m2+motor.m3+motor.m4)/4 should equal thrust. However, upon examining the CSV values below, I confirmed that this is not the case.

[motor_data.csv](https://github.com/bitcraze/discussions/files/15195728/motor_data.csv)

Q1. According to the Motor Power Distribution's Legacy Mode, the value obtained by adding the motors and dividing by 4 should equal thrust, but it doesn't. Why is that?

Q2. I want to have the same values for Thrust and each motor's PWM. To achieve this, do I simply need to comment out the roll, pitch, and yaw values at https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L85-L88

**Comment by ataffanel** [comment]
I think your problem is because you are not taking into account the full chain. The thrust is not only calculated from the power distribution but also passes through other functions to de-saturate the output (ie. avoid roll/pitch/yaw unbalance in case the thrust if too high). The full chain that generated the thrust is in the stabilizer loop: https://github.com/bitcraze/crazyflie-firmware/blob/d9ee34a78cecf2eaf8b05e0b54487cff0c7b4590/src/modules/src/stabilizer.c#L249-L255.

**Reply by WoodyloveBo** [reply]
Thank you. Thanks to you, I was able to understand the whole chain.
As mentioned in Q2 earlier, can I make the PWM of each motor have the same value by commenting out the roll, pitch, and yaw variables?

https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/power_distribution_quadrotor.c#L85-L88 ?

**Reply by gemenerik** [reply, marked as answer]
Yes you can do something along those lines, but obviously you are going to lose you ability to adjust the attitude of the drone and it will crash, so be warned.
